File: src/main/java/dev/goerner/geozen/jackson/deserializer/PointDeserializer.java
package dev.goerner.geozen.jackson.deserializer;

import dev.goerner.geozen.model.simple_geometry.Point;
import tools.jackson.core.JsonParser;
import tools.jackson.databind.DeserializationContext;
import tools.jackson.databind.JsonNode;

public class PointDeserializer extends AbstractGeometryDeserializer<Point> {

    @Override
    public Point deserialize(JsonParser p, DeserializationContext ctxt) {
        JsonNode rootNode = p.readValueAsTree();

        checkType(rootNode, "Point");

        return new Point(parsePosition(rootNode.get("coordinates")));
    }
}

File: src/main/kotlin/dev/goerner/geozen/jackson/deserializer/PointDeserializer.kt
package dev.goerner.geozen.jackson.deserializer

import dev.goerner.geozen.model.simple_geometry.Point
import tools.jackson.core.JsonParser
import tools.jackson.databind.DeserializationContext
import tools.jackson.databind.JsonNode

class PointDeserializer : AbstractGeometryDeserializer<Point>() {

    override fun deserialize(p: JsonParser, ctxt: DeserializationContext): Point {
        val rootNode = p.readValueAsTree<JsonNode>()

        checkType(rootNode, "Point")

        val coordinates = rootNode["coordinates"]
        require(coordinates != null) { "Missing 'coordinates' field" }

        return Point(parsePosition(coordinates))
    }
}
